dev-guide/feature-gate.md

```diff
@@ -0,0 +1,293 @@
+
+	// possible (probable?) future additions include
+	// 1. support level (Stable, ServiceDeliveryOnly, TechPreview, DevPreview)
+	// 2. description
+}
+```
+
+#### openshift/library-go
+The existing config observer will be updated to be a drop-in replacement (this covers a dozen or so operators).
+Operators that don't use that pattern can use one like this
+
+```go
+featureGateAccessor := NewFeatureGateAccess(args)
+go featureGateAccessor.Run(ctx)
+
+// wait for feature gates to observed.
+select{
+case <- featureGateAccessor.InitialFeatureGatesObserved():
+	enabled, disabled, _ := featureGateAccessor.CurrentFeatureGates()
+	klog.Infof("FeatureGates initialized: enabled=%v  disabled=%v", enabled, disabled)
+case <- time.After(1*time.Minute):
+	klog.Errorf("timed out waiting for FeatureGate detection")
+	return fmt.Errorf("timed out waiting for FeatureGate detection")
+}
+
+// continue from here.
+// also open to changes to wire up to k/k FeatureGates if there is demand.
+```
+
+library-go also has code allow a "default" payload version (one that hasn't been substituted) to mean, "give me the 
+feature gates for the latest CVO version".
+This will support the development patch use-cases.
+
+### Workflow Description
+
+#### Variation [optional]
+
+### API Extensions
+
+No admission webhooks are needed.
+If complicated validation is required, there is already a FeatureGate admission plugin hardcoded in our openshift/kubernetes
+fork and we'll use that.
+
+### Implementation Details/Notes/Constraints [optional]
+
+### Risks and Mitigations
+
+### Drawbacks
+
+## Design Details
+
+### Open Questions [optional]
+
+### Test Plan
+
+Unit tests are already added to implementation PRs.
+Clusters will not install without the controller functional and we don't support changing feature gates back from
+TechPreviewNoUpgrade, so e2e tests are impractical.
+
+### Graduation Criteria
+
+#### Dev Preview -> Tech Preview
+
+#### Tech Preview -> GA
+
+Since this is actually controlling the gates, it is not practical to pre-test it.
+Once unit tests pass and clusters successfully install equivalently using this mechanism, PRs will be opened against 
+several initial operators, including
+1. etcd
+2. kube-apiserver
+3. kube-controller-manager
+4. kube-scheduler
+5. openshift-apiserver
+6. openshift-controller-manager
+7. authentication
+8. ingress
+9. cloud-credential-operator
+10. networking
+11. storage
+
+There is incentive to use this mechanism when introducing TechPreviewNoUpgrade features because consuming feature gates
+this way is less work.
+
+We're done when all operators using feature gates are reading from FeatureGate.Status instead of vendoring the list.
+
+#### Removing a deprecated feature
+
+- Announce deprecation and support policy of the existing feature
+- Deprecate the feature
+
+### Upgrade / Downgrade Strategy
+
+On upgrade to the first level with this change, the cluster-config-operator goes first, so the FeatureGateStatus will
```

Comment:
Is this true? I just double checked and the CCO deployment manifest is at run level 30, so I wouldn't expect it to be upgraded until after all of the core control plane components